[GSoC] Function for plotting loading results added

Psycho-Pirate opened the pull request:
<!-- Your title above should be a short description of what
was changed. Do not include the issue number in the title. -->

#### References to other Issues or PRs
<!-- If this pull request fixes an issue, write "Fixes #NNNN" in that exact
format, e.g. "Fixes #1234" (see
https://tinyurl.com/auto-closing for more information). Also, please
write a comment on that issue linking back to this pull request once it is
open. -->
See #19202

#### Brief description of what is fixed or changed
A function for plotting all loading results is added in Beam3D class.

#### Other comments


#### Release Notes



<!-- Write the release notes for this release below between the BEGIN and END
statements. The basic format is a bulleted list with the name of the subpackage
and the release note for this PR. For example:

* solvers
  * Added a new solver for logarithmic equations.

* functions
  * Fixed a bug with log of integers.

or if no release note(s) should be included use:

NO ENTRY

See https://github.com/sympy/sympy/wiki/Writing-Release-Notes for more
information on how to write release notes. The bot will check your release
notes automatically to see if they are formatted correctly. -->

<!-- BEGIN RELEASE NOTES -->

* physics.continuum_mechanics
  * Added ability to plot all loading results in Beam3D class

<!-- END RELEASE NOTES -->

Psycho-Pirate committed:
function for plotting loading results added

Psycho-Pirate commented:
Here is an example:
```
>>> from sympy.physics.continuum_mechanics.beam import Beam3D
            >>> from sympy import symbols
            >>> l, E, G, I, A, x = symbols('l, E, G, I, A, x')
            >>> b = Beam3D(20, 40, 21, 100, 25, x)
            >>> b.apply_load(15, start=0, order=0, dir="z")
            >>> b.apply_load(12*x, start=0, order=0, dir="y")
            >>> b.bc_deflection = [(0, [0, 0, 0]), (20, [0, 0, 0])]
            >>> R1, R2, R3, R4 = symbols('R1, R2, R3, R4')
            >>> b.apply_load(R1, start=0, order=-1, dir="z")
            >>> b.apply_load(R2, start=20, order=-1, dir="z")
            >>> b.apply_load(R3, start=0, order=-1, dir="y")
            >>> b.apply_load(R4, start=20, order=-1, dir="y")
            >>> b.solve_for_reaction_loads(R1, R2, R3, R4)
            >>> b.solve_slope_deflection()
            >>> b.plot_loading_results('y')
```
![all](https://user-images.githubusercontent.com/50932406/122943716-e1180380-d394-11eb-8229-1ef197667bea.png)

Psycho-Pirate commented:
@ishanaj @Sc0rpi0n101

Psycho-Pirate commented:
I tried to build the documentation but I am getting a weird order of functions in the final docs.
![docss](https://user-images.githubusercontent.com/50932406/122944394-6b606780-d395-11eb-8e41-5d19eeeb1ea3.png)
Here ```plot_loading_results()``` is shown before ```plot_shear_force()``` even though I wrote them in the correct order in beam.py

ishanaj commented:
> Here plot_loading_results() is shown before plot_shear_force() even though I wrote them in the correct order in beam.py

I think it comes up in alphabetical order. No issues!

ishanaj commented:
Could you try a 4x3 grid for the `dir="all"` case and share the results here?

ishanaj commented on sympy/physics/continuum_mechanics/beam.py:
```diff
@@ -2667,3 +2667,104 @@ def plot_deflection(self, dir="all", subs=None):
+
+        """
+
+        if dir == None:
```
Should rather be `if not dir:`

ishanaj commented on sympy/physics/continuum_mechanics/beam.py:
```diff
@@ -2667,3 +2667,104 @@ def plot_deflection(self, dir="all", subs=None):
+        if subs is None:
+            subs = {}
+
+        for sym in deflection[dir_num].atoms(Symbol):
```
I think this should be done for slope, shear force, and bending moment also. Better to directly use the helper functions of the individual plotting functions (`_plot_slope() ..etc`)

Psycho-Pirate replied:
I did it for deflection only because deflection covers all the variables needed for calculating slope, bending moment, and shear force. If deflection is determinable, we can definitely determine other equations as the code works in order.

I didn't use the helper functions because I wanted to impart different colors to different functions. Using helper functions will plot all the x graphs as red, y as blue, etc unless we modify the helper functions to take color as a parameter.

ishanaj replied:
I think the idea is to have different line colors within a single plot. So you can change the color of the helper functions in a way that when you use them here all the line colors in this plot are different. (I won't recommend using color as a parameter)

ishanaj replied:
> ....because deflection covers all the variables needed for calculating slope, bending moment, and shear force

IMO there can be a use case where this won't hold true. Or is this something proven?

Psycho-Pirate replied:
> IMO there can be a use case where this won't hold true. Or is this something proven?

The functions are calculated by integrating the lower functions starting from the load. Shear Force is calculated by integrating load, Bending Moment is calculated by integrating shear force, and so on. Deflection is the function at the top so it covers everything below it and adds on some additional variables. Hence it contains all the variables needed for all the other functions. If Deflection can be calculated, it can be proven that other functions have already been calculated and the variables they contain are already present. 

Even in beam2d class, only Deflection is used for ```plot_loading_results()``` because of the same reason.

Psycho-Pirate commented:
> Could you try a 4x3 grid for the `dir="all"` case and share the results here?

If I do not specify a custom size, it looks like this
![loadingres2](https://user-images.githubusercontent.com/50932406/123280741-a8a82f00-d526-11eb-94cb-cc66dfce3988.png)

But if I define a size of 20x10 for each subplot, it looks like this
![loadingres](https://user-images.githubusercontent.com/50932406/123280856-c83f5780-d526-11eb-895b-1864c907331d.png)
I am not sure how it will look in the documentation with the custom size, I think it will be pretty small for the user to get any valuable information.

Psycho-Pirate committed:
changes made

ishanaj commented:
Also, the default `dir` of `plot_loading_results()` can be made `x`, possibly because the same function in 2d doesn't take any arguments, so I think this should also work without any arguments(not considering `subs` here).

Psycho-Pirate committed:
changes made

Psycho-Pirate committed:
changes made

Psycho-Pirate committed:
changes made

Psycho-Pirate commented:
In the latest commit  I have made the following changes:
* changed colors for every plot
* set default direction to x in ```plot_loading_results()```
* used helper functions in ```plot_loading_results()```
* removed the ```if sym == self.variable: continue``` condition
* used subs in the example

@ishanaj @Sc0rpi0n101

ishanaj commented:
LGTM
